builder/azure/arm/step_deploy_template.go

```diff
@@ -279,82 +324,141 @@ func (s *StepDeployTemplate) deleteImage(ctx context.Context, imageName string,
+func (s *StepDeployTemplate) deleteNetworkInterface(ctx context.Context, id commonids.NetworkInterfaceId) error {
+	pollingContext, cancel := context.WithTimeout(ctx, s.client.PollingDuration)
+	defer cancel()
+	err := s.retryDeletion(pollingContext, id.NetworkInterfaceName, NetworkInterfaceResourceType, func() error {
+		return s.client.NetworkMetaClient.NetworkInterfaces.DeleteThenPoll(pollingContext, id)
+	})
+	return err
+}
+
+func (s *StepDeployTemplate) listDeploymentOperations(ctx context.Context, id deploymentoperations.ResourceGroupDeploymentId) ([]deploymentoperations.DeploymentOperation, error) {
 	var maxResources int64 = 50
 	options := deploymentoperations.DefaultListOperationOptions()
 	options.Top = &maxResources
-	id := deploymentoperations.NewResourceGroupDeploymentID(subscriptionId, resourceGroupName, deploymentName)
 	pollingContext, cancel := context.WithTimeout(ctx, s.client.PollingDuration)
 	defer cancel()
 
 	deploymentOperations, err := s.client.DeploymentOperationsClient.ListComplete(pollingContext, id, options)
 	if err != nil {
-		s.reportIfError(err, resourceGroupName)
-		return err
+		return nil, err
 	}
+	return deploymentOperations.Items, nil
+}
 
-	resources := map[string]string{}
-
-	for _, deploymentOperation := range deploymentOperations.Items {
-		// Sometimes an empty operation is added to the list by Azure
-		if deploymentOperation.Properties.TargetResource == nil {
-			continue
-		}
-
-		resourceName := *deploymentOperation.Properties.TargetResource.ResourceName
-		resourceType := *deploymentOperation.Properties.TargetResource.ResourceType
-
-		s.say(fmt.Sprintf("Adding to deletion queue -> %s : '%s'", resourceType, resourceName))
-		resources[resourceType] = resourceName
-
+// This function is called to delete the resources remaining in the deployment after we delete the Virtual Machine and the deleteNic
+// These resources before the VM and the NIC results in errors
+func (s *StepDeployTemplate) deleteDetatchedResourcesWithQueue(ctx context.Context, subscriptionId string, resourceGroupName string, resources map[string]string) error {
+	retryConfig := retry.Config{
+		Tries: 5,
+		RetryDelay: (&retry.Backoff{
+			InitialBackoff: 3 * time.Second,
+			MaxBackoff:     15 * time.Second,
+			Multiplier:     1.5,
+		}).Linear,
 	}
-
 	var wg sync.WaitGroup
 	wg.Add(len(resources))
 
 	for resourceType, resourceName := range resources {
 		go func(resourceType, resourceName string) {
 			defer wg.Done()
-			retryConfig := retry.Config{
-				Tries:      10,
-				RetryDelay: (&retry.Backoff{InitialBackoff: 5 * time.Second, MaxBackoff: 60 * time.Second, Multiplier: 1.5}).Linear,
-			}
 
-			err = retryConfig.Run(ctx, func(ctx context.Context) error {
-				s.say(fmt.Sprintf("Attempting deletion -> %s : '%s'", resourceType, resourceName))
+			err := retryConfig.Run(ctx, func(ctx context.Context) error {
```

Comment:
Any reason why we don't use `retryDeletion` for those resources?

Reply:
Nope, good call out, I moved this to using that!